src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -20952,32 +20952,409 @@ public class C
+    public string field3;
+    public string? field4;
+}
+public class C : Base
+{
+    public C()
+    {
+        init();
+        field1.ToString();
+        field2.ToString();
+        field3.ToString();
+        field4.ToString();
+
+        [MemberNotNull(nameof(field1), nameof(field2))]
+        void init() => throw null!;
+    }
+}
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyDiagnostics(
+                // (4,19): warning CS8618: Non-nullable field 'field1' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the field as nullable.
+                //     public string field1;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "field1").WithArguments("field", "field1").WithLocation(4, 19),
+                // (6,19): warning CS8618: Non-nullable field 'field3' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the field as nullable.
+                //     public string field3;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "field3").WithArguments("field", "field3").WithLocation(6, 19),
+                // (15,9): warning CS8602: Dereference of a possibly null reference.
+                //         field2.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field2").WithLocation(15, 9),
+                // (17,9): warning CS8602: Dereference of a possibly null reference.
+                //         field4.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(17, 9),
+                // (19,10): warning CS8776: Member 'field1' cannot be used in this attribute.
+                //         [MemberNotNull(nameof(field1), nameof(field2))]
+                Diagnostic(ErrorCode.WRN_MemberNotNullBadMember, "MemberNotNull(nameof(field1), nameof(field2))").WithArguments("field1").WithLocation(19, 10),
+                // (19,10): warning CS8776: Member 'field2' cannot be used in this attribute.
+                //         [MemberNotNull(nameof(field1), nameof(field2))]
+                Diagnostic(ErrorCode.WRN_MemberNotNullBadMember, "MemberNotNull(nameof(field1), nameof(field2))").WithArguments("field2").WithLocation(19, 10));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_LocalFunction_WithinInstanceMethod()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1 = "";
+    public string? field2;
+    public string field3 = "";
+    public string? field4;
+
+    public void M()
+    {
+        init();
+        field1.ToString();
+        field2.ToString();
+        field3.ToString();
+        field4.ToString(); // 1
+
+        [MemberNotNull(nameof(field1), nameof(field2))]
+        bool init() => throw null!;
```

Comment:
>bool init() => throw null!; [](http://example.com/codeflow?start=8&length=27)

Consider testing the equivalent case with a `static` local function, with the same instance members.
```csharp
[MemberNotNull(nameof(field1), nameof(field2))]
static bool init() => throw null!;
```